flowapi/flowapi/permissions.py

```diff
@@ -267,86 +154,98 @@ def schema_to_scopes(schema: dict) -> Iterable[str]:
+                    "oneOf": [{"$ref": "DUMMY"}]},
+                "DUMMY": {"properties": {"query_kind": {"enum": ["dummy"]}}},
+            },
+        )
+    )
 
-    Parameters
-    ----------
-    scopes : list of str
-        Compressed scopes to expand
+    ["dummy", "run&dummy", "get_result&available_dates"],
+    """
+    resolved_queries = ResolvingParser(spec_string=dumps(schema)).specification[
+        "components"
+    ]["schemas"]["FlowmachineQuerySchema"]
+    unique_scopes = []
+    for tl_query in resolved_queries["oneOf"]:
+        tl_query_name = tl_query["properties"]["query_kind"]["enum"][0]
+        print(f"Looking for {tl_query_name}")
+        query_list = grab_on_key_list(
+            tl_query,
+            ["properties", "query_kind", "enum", 0],
+        )
+        if query_list == []:
+            return []
+        scopes_generator = (
+            tl_schema_scope_string(tl_query, query) for query in query_list
+        )
+        unique_scopes += list(set.union(*scopes_generator))
+    return sorted(unique_scopes)
 
-    Yields
-    ------
-    str
-        A scope string
 
+async def get_agg_unit(query_dict):
     """
-    for scope in scopes:
-        parts = scope.strip().split("&")
-        ps = (x.split(",") for x in parts)
-        yield from (set(x) for x in product(*ps))
-
-
-@functools.singledispatch
-def query_to_scope_list(tree, paths=None, keep=["aggregation_unit"]) -> str:
+    Interrogates Flowmachine for the top-level agg unit of this query
     """
-
-    Parameters
-    ----------
-    tree : list of dict or dict
-    paths : tuple
-    keep : list of str
-        List of fields to include in scope strings
-
-    Yields
-    ------
-    str
-        Scope string
+    request.socket.send_json(
+        {
+            "request_id": request.request_id,
+            "action": "get_aggregation_unit",
+            "params": query_dict,
+        }
+    )
+    reply = await request.socket.recv_json()
+    if reply["status"] != "success":
+        raise Exception("Query has no type - something wrong with Flowmachine")
+    try:
+        return reply["payload"]["aggregation_unit"]
+    except KeyError:
+        raise Exception(
+            "Reply missing aggregation_unit key - something wrong with Flowmachine."
+            f"Received output: {reply}"
+        )
 
 
+async def query_to_scopes(query_dict):
     """
-    yield from ()
-
-
-@query_to_scope_list.register
-def _(tree: list, paths=None, keep=["aggregation_unit"]) -> str:
-    if paths is None:
-        paths = tuple()
-    for v in tree:
-        yield from query_to_scope_list(v, paths, keep=keep)
+    Given a query_dict of the form
+    {
+        query_kind:tl_query,
+        aggregation_unit:agg_unit
+        ...
+        sub_param:{
+            query_kind: sub_query...}
+    }
+    returns the scope triplets of the query in the form "agg_unit:tl_query:sub_query".
+    Will always return "agg_unit:tl_query:tl_query"
```

Comment:
I wonder whether it would be best to only return `agg_unit:tl_query:tl_query` if there are no sub-queries? (I.e. if a user has the required sub-query scopes, they don't need to also have the additional top-level scope). Effectively this would change the meaning of `agg_unit:tl_query:tl_query` from "can run a `tl_query` query" to "can run a `tl_query` query with no sub-queries".

Otherwise a role with scope `admin3:spatial_aggregate:daily_location` but not `admin3:spatial_aggregate:spatial_aggregate` would not be able to run any spatial aggregate query, but it's not necessarily clear to somebody creating a role in FlowAuth that they have to tick the `admin3:spatial_aggregate:spatial_aggregate` box otherwise any other selected `admin3:spatial_aggregate:X` scopes are meaningless.

An alternative could be to enforce this on the FlowAuth side (i.e. if `agg_unit:tl_query:sub_query` is selected, automatically add `agg_unit:tl_query:tl_query` to the role as well). But I think overall it's cleaner to handle this in FlowAPI.